cassiopeia/solver/dissimilarity_functions.py

```diff
@@ -225,3 +221,95 @@ def weighted_hamming_similarity(
+    |a_i| is the number of ambiguous characters at target site i of string a,
+    and |b_i| is the number of amiguous characters at target site i of string b.
+    As an example, if we have N=10 and all a_i=b_i=2, then we have to construct
+    1,038,576 * 2 strings and compute over 4 trillion dissimilarities.
+
+    By assuming each target site is independent, simply calculating the sum of
+    the linkages of each target site separately is equivalent to the naive
+    implementation (can be proven by induction). This procedure is implemented
+    in this function. One caveat is that we usually normalize the distance by
+    the number of shared non-missing positions. We approximate this by dividing
+    the absolute distance by the sum of the probability of each site not being
+    a missing site for both strings.
+
+    The idea of linkage is analogous to that in hierarchical clustering, where
+    ``np.min`` can be used for single linkage, ``np.max`` for complete linkage,
+    and ``np.mean`` for average linkage (the default).
+
+    The reason the ``dissimilarity_function`` argument is defined as the first
+    argument is so that this function may be used as input to
+    :func:`cassiopeia.data.CassiopeiaTree.compute_dissimilarity_map`. This can
+    be done by partial application of this function with the desired dissimilarity
+    function.
+
+    Note:
+        If neither character string is ambiguous, then calling this function is
+        equivalent to calling ``dissimilarity_function`` separately.
+
+    Args:
+        s1: The first (possibly) ambiguous sample
+        s2: The second (possibly) ambiguous sample
+        missing_state_indicator: The character representing missing values
+        weights: A set of optional weights to weight the similarity of a mutation
+        dissimilarity_function: The dissimilarity function to use to calculate pairwise
+            dissimilarities.
+        linkage_function: The linkage function to use to aggregate dissimilarities
+            into a single number. Defaults to ``np.mean`` for average linkage.
+        normalize: Whether to normalize to the proportion of sites present in
+            both strings.
+
+    Returns:
+        The dissimilarity between the two ambiguous samples
+    """
+    # Make any unambiguous character strings into pseudo-ambiguous so that we
+    # can easily use itertools.product
+    s1 = [s if isinstance(s, tuple) else (s,) for s in s1]
+    s2 = [s if isinstance(s, tuple) else (s,) for s in s2]
+
+    result = 0
+    num_present = 0
+    for i, (c1, c2) in enumerate(zip(s1, s2)):
+        dissim = []
+        present = []
+        for _c1, _c2 in itertools.product(c1, c2):
+            present.append(
+                _c1 != missing_state_indicator
+                and _c2 != missing_state_indicator
+            )
+            dissim.append(
+                dissimilarity_function(
+                    [_c1],
+                    [_c2],
```

Comment:
If you're using one of the numba optimized functions, they might start complaining that these aren't numba lists or numpy arrays -- can you wrap these with np.array? Or am I incorrect?

Reply:
The `dissimilarity_function` is not numbaized because the overhead due to calling the numbaized function multiple times (with very small arguments) overweighs any speedup (around 100x slower).

Reply:
Perhaps I'm misunderstanding something, but isn't that precisely when Numba should be given us a speedup? I understand that if you need to recompile a Numba function several times it might slow us down, but if you only compile once then the function should speed things up if you call it several times, right?